[WIP] Bring music services back

KennethNielsen opened the pull request:
This PR contains the WIP for bringing music services back. I've been working of different sources for this and will of course credit them before merging.

**NEEDS TESTING**

The short history of music service authentication with Sonos is that Sonos used to store and make available _account_ information (which covered the main access to a music service), which meant that the only thing we had to do was to query the speaker for all the information we needed to talk to the music services. Then at some point, some guess due to pressure from the music services, this was changed. One of the mechanisms which replaced this was different forms of browser authenticated token access, which involves authenticating in the browser, to get an access token. Two of these methods are called DeviceLink and AppLink and it is these that this PR aims to implement.

The code has only so far been tested successfully on one music service, Tidal, which uses DeviceLink, so I am in need of lots more testing and especially on an AppLink music service, such as Spotify.

You can test the new auth features with code like this:
```python
from soco.music_services import MusicService
music_service = MusicService("Spotify")
reg_url, link_code = music_service.device_or_app_link_auth_part1()
print(reg_url)
print(link_code)
```

Visit the url and authenticate and store the link_code for later. Then, after having authenticated, call part 2, with the link code as only argument
```python
music_service.device_or_app_link_auth_part2("THE_LINK_CODE_FROM_BEFORE")
```
and then, if everything goes well, we should be authenticated. **It is very important to not call part1 again before when calling part 2, since then it will be a new link code, a wrong one. The link code passed into part2 has to be the one returned along with the reg URL actually used**. The tokens by default are stored in a json file in the users default config dir so e.g. on Linux you can see the stored token with the command:
```bash
cat ~/.config/SoCo/token_store.json
```

Now the music service should be available for use like before, where items can be browsed and searched for and the results enqueued. You can use code along the lines of this to quickly test it (only editing the name of the speaker away from "Kontor" and specify the name of the music service at the bottom:

```python
from time import sleep
from random import choice
import soco
from soco.music_services import MusicService
from soco.discovery import by_name

def confirm_work(service):
    t = MusicService(service)
    z = by_name("Kontor")

    # Print the search categories. TIDAL e.g. has 4 search categories:
    # ['artists', 'albums', 'tracks', 'playlists']
    print(t.available_search_categories)

    for search_category in t.available_search_categories:
        z.clear_queue()
        sleep(1)
        # Clear queue
        print("Search category:", search_category)
        results = t.search(search_category, "Metallica")
        print(len(results))
        try:
            z.add_to_queue(choice(results))
        except soco.exceptions.SoCoUPnPException:
            print("Cannot play")
        else:
            print("Can play")
            print(str(z.get_queue())[:120])
            z.play()
            sleep(10)
            z.stop()

    root_content = t.get_metadata()
    print("\nGot {} items in root content".format(len(root_content)))
    print("\nTest browse root items:")
    for item in root_content:
        content = t.get_metadata(item.id)
        print("Got {} items under: {}".format(len(content), item.title))

    print("\nTest browse until tracks")
    for item in t.get_metadata():
        # print("J")
        while True:
            print("-> (" + type(item).__name__ + ")", item.title)
            items = t.get_metadata(item.id)
            if len(items) == 0 or items[0].id == item.id:
                print("... STOP\n")
                break
            item = items[0]

    return

confirm_work("Spotify")
```

I'm looking forward to hearing how it is working out for you. Please post testing output from the code above if available along.

BTW: Never mind the code (which is a little messy for now) or the tests failing (which I haven't revised), this is just for getting the code tested on new music services.

KennethNielsen commented:
If you are in doubt about how you music service is spelled, just ask the MusicService class:
```python
print(MusicService.get_all_music_services_names())
```

amelchio commented:
It mostly works with Spotify.

I had to use `device_link_auth_part1()`, not `device_or_app_link_auth_part1()`. It then failed to extract `regUrl` and `linkCode`, my result was formatted like this:

```
OrderedDict([('authorizeAccount',
              OrderedDict([('appUrlStringId', 'SIGN_IN'),
                           ('deviceLink',
                            OrderedDict([('regUrl',
                                          'https://spotify-v5.ws.sonos.com/deviceLink/home?linkCode=EFCYTA'),
                                         ('linkCode', 'EFCYTA'),
                                         ('showLinkCode', 'false')]))])),
             ('createAccount',
              OrderedDict([('appUrlStringId', 'CREATE_NEW')]))])
```

Token store on macOS seems to be `~/Library/Application\ Support/SoCo/token_store.json`.

Finally, I had to add `from soco.discovery import by_name` to your test program. I did _not_ have to change "Kontor" though :-)

The program produced this output:
```
$ python3 kenneth.py
[]

Got 5 items in root content

Test browse root items:
Got 14 items under: Musik til søndagen
Got 13 items under: Charts
Got 50 items under: New Releases
Got 42 items under: Genres and Moods
Got 3 items under: Your Music

Test browse until tracks
-> (MSContainer) Musik til søndagen
-> (MSPlaylist) Sunday Funday
-> (MSTrack) Summer Nights
... STOP

-> (MSCollection) Charts
-> (MSPlaylist) It's Hits Denmark
-> (MSTrack) Monster (Shawn Mendes & Justin Bieber)
... STOP

-> (MSAlbumlist) New Releases
-> (MSAlbum) Mindre
-> (MSTrack) Mindre
... STOP

-> (MSCollection) Genres and Moods
-> (MSCollection) Top Lists
-> (MSPlaylist) It's Hits Denmark
-> (MSTrack) Monster (Shawn Mendes & Justin Bieber)
... STOP

-> (MSCollection) Your Music
-> (MSCollection) Playlists
-> (MSPlaylist) Daily Mix 5
-> (MSTrack) Everybody's Gonna Be Happy - Non-LP Version
... STOP
```

KennethNielsen commented:
@amelchio thank you so much.

I'm sorry about the naming of the methods and that extracting didn't work. I forgot to push the last commit which takes care of that :$ . Actually one of the remaining question I have, is whether these return structures are consistently structured, but I actually developed the part1 part for AppLink for Spotify (which I can do without an account), so it should now work smoothly.

Any chance I could persuade you to:
1) Delete the token you got and test again that it goes smooth and without problems at 1b1c208
2) Re-run the test. The beginning part of the test should actually test playing something on your speaker, but it might have been disabled before I edited it and fixed or it might actually have been there, in which case I'm interested in whether it worked.

I also fixed the missing import in the test code.

amelchio commented:
That's better but playing does not work with the test program, there is no `PresentationMap` for Spotify. However, there is this:

```
<Manifest Version="8" Uri="https://mnfts.ws.sonos.com/p/spotify-manifest.json"/>
```

and you can see that this URL contains a `presentationMap`. So I got it working with these changes: https://github.com/amelchio/pysonos/commit/981476071823d1e474c86eac33681391aecd6993 ... I also added an `import soco` in the test program, for the exception.

KennethNielsen commented:
@amelchio thanks, patch applied and import fixed.If I could persuade you to test again, that would be super

pwt commented:
The first couple of times I ran the setup code, the part 2 auth failed with a SOAP error. No `token_store.json` file had been created. The third time it worked. Perhaps this is a timing thing.

Running the test code then produces the following output:

```
(venv) pwt-mba:SoCo pwt$ python test_spotify.py 
['artists', 'albums', 'tracks', 'playlists']
Search category: artists
7
Cannot play
Search category: albums
50
Cannot play
Search category: tracks
50
Cannot play
Search category: playlists
50
Cannot play

Got 5 items in root content

Test browse root items:
Got 10 items under: Your afternoon listening.. sorted
Got 14 items under: Charts
Got 50 items under: New Releases
Got 50 items under: Genres and Moods
Got 3 items under: Your Music

Test browse until tracks
-> (MSContainer) Your afternoon listening.. sorted
-> (MSPlaylist) The Rock List
-> (MSTrack) Animals
... STOP

-> (MSCollection) Charts
-> (MSPlaylist) Hot Hits UK
-> (MSTrack) positions
... STOP

-> (MSAlbumlist) New Releases
-> (MSAlbum) Plastic Hearts
-> (MSTrack) WTF Do I Know
... STOP

-> (MSCollection) Genres and Moods
-> (MSCollection) Top Lists
-> (MSPlaylist) Hot Hits UK
-> (MSTrack) positions
... STOP

-> (MSCollection) Your Music
-> (MSCollection) Playlists
-> (MSPlaylist) Chilled
-> (MSTrack) Sweet Harmony
... STOP
```

A subsequent run gave the following error:

`soco.soap.SoapFault: ns8eeLx:Client.AuthTokenExpired: authTokenExpired`

goldyfruit commented:
@KennethNielsen I was able run the authentication, the `token_store.json` has been created with the token.

Here is the test script output:

```
['artists', 'albums', 'tracks', 'playlists']
Search category: artists
7
Cannot play
Search category: albums
50
Can play
Queue(items=[<DidlMusicTrack 'b'Metallica - Freestyle'' at 0x7f8e26e3c8>])
Search category: tracks
50
Can play
Queue(items=[<DidlMusicTrack 'b'Metallica'' at 0x7f8e2ea2b0>])
Search category: playlists
50
Can play
Queue(items=[<DidlMusicTrack 'b'Leper Messiah (Remastered)'' at 0x7f8e1faa20>, <DidlMusicTrack 'b'Welcome Home (Sanitari

Got 5 items in root content

Test browse root items:
Got 10 items under: Let the music play all night long.
Got 22 items under: Charts
Got 50 items under: New Releases
Got 50 items under: Genres and Moods
Got 3 items under: Your Music

Test browse until tracks
-> (MSContainer) Let the music play all night long.
-> (MSPlaylist) Northern Bars
-> (MSTrack) No Brakes
... STOP

-> (MSCollection) Charts
-> (MSPlaylist) Hot Hits Canada
-> (MSTrack) YO VISTO ASÍ
... STOP

-> (MSAlbumlist) New Releases
-> (MSAlbum) EL ÚLTIMO TOUR DEL MUNDO
-> (MSTrack) EL MUNDO ES MÍO
... STOP

-> (MSCollection) Genres and Moods
-> (MSCollection) Top Lists
-> (MSPlaylist) Hot Hits Canada
-> (MSTrack) YO VISTO ASÍ
... STOP

-> (MSCollection) Your Music
-> (MSCollection) Playlists
-> (MSPlaylist) SoundHound
-> (MSTrack) May I Have This Dance (Remix) [feat. Chance the Rapper]
... STOP
```

goldyfruit commented:
One thing to mention is that I had a Python dependency issue, I had to install `appdirs` .

```
$ pip install appdirs
```

I guess this will be added to the `requirements.txt` later.

KennethNielsen commented:
@pwt thanks for testing. I'm quite interested in your results, because they are different from @goldyfruit for the same service. Especially I find it troubling that the first part of the test, which is meant to test if the data structures you receive can actually be played, and they seem to all fail. So I have a couple of questions:
1) When you mentioned that auth failed on the first few attempts, are you 100% sure you followed procedure, in that you stored the link code from part1 and used that in part 2 and did the browser auth in between? I know you probably did, just checking.
2) I'm quite interested in that fact that the play test failed and in whether that is a problem with the test or the music service. Can I persuade you to test separately, whether you can add e.g. a track that you got from the Spotify service to the queue and play it?

Finally. I'm real interested in that fact that you eventually got the AuthTokenExpired, because my Tidal token hasn't expired yet and I'm still missing to implement the automatic renewal. EDIT. It just did. I have pushed a new commit which tries to handle the token refresh by parsing the new private key and token passed along with the error when the old one has expired. Can you test if this works for you. I.e. pull the new commit, run any call on the music service requiring auth (e.g. search for a track) and check your tokenstore before and after to see if it was properly updated.

KennethNielsen commented:
@goldyfruit thanks for the testing and great that it works. WRT to appdirs, there is just a lot of cleanup to do still, but I added appdirs now.

KennethNielsen commented:
OK everyone. I have given the token-refresh implementation a stab. The tokens we get are temporary, and needs to be refreshed after a while. The new token is however returned along with the error that the old one has expired, so it is just about extracting it. I have tried to do this now based on Tidal data, but only partially tested it*, so it needs tests. So if you tested the first part and it has been a few days, most likely your token will have expired, and you can test this. The code is still fairly verbose, with lots of prints in there, so there should be a repeated RE-AUTH line if it actually reached that part of the code. If it fails, please paste all output, since there should be a print out of the XML in there, which I can have a look at.

* I got a new token and the xml parsing failed, so I fixed the parsing so it works in a REPL and then updated the code, but I that point I had just gotten new tokens, so I could not force a test of the new code

pwt commented:
> 1. When you mentioned that auth failed on the first few attempts, are you 100% sure you followed procedure, in that you stored the link code from part1 and used that in part 2 and did the browser auth in between? I know you probably did, just checking.

I believe I followed the required steps. (But, hey ...)

> 2. I'm quite interested in that fact that the play test failed and in whether that is a problem with the test or the music service. Can I persuade you to test separately, whether you can add e.g. a track that you got from the Spotify service to the queue and play it?

Sure. I'll get some time later today (or tomorrow) to do some more testing.

> Finally. I'm real interested in that fact that you eventually got the AuthTokenExpired, because my Tidal token hasn't expired yet and I'm still missing to implement the automatic renewal. EDIT. It just did. I have pushed a new commit which tries to handle the token refresh by parsing the new private key and token passed along with the error when the old one has expired. Can you test if this works for you. I.e. pull the new commit, run any call on the music service requiring auth (e.g. search for a track) and check your tokenstore before and after to see if it was properly updated.

Will do. Very exciting to see the progress on this.

goldyfruit commented:
@KennethNielsen I'm not sure about the steps but I did pull you changes, didn't run the `auth1` et `auth2` but I got `tokenRefreshRequired`.

My `~/.config/SoCo/token_store.json` still there with my previous token.

This is the code I tried to run without the token refresh.
```
import soco
from soco.music_services import MusicService
from soco.discovery import by_name
from random import choice

service = MusicService("Spotify")
speaker = by_name("Salon")
playlist = choice(playlists)
speaker.add_to_queue(playlist)
print(playlist)
```

Output:
```
SoapFault
ns8eeLx:Client.TokenRefreshRequired
<Element 'detail' at 0x7f95e53a98>
Traceback (most recent call last):
  File "test.py", line 16, in <module>
    playlists = service.search("playlists", "christmas classic")
  File "/home/pi/venvs/smartgic-api/lib/python3.7/site-packages/soco/music_services/music_service.py", line 894, in search
    ("count", count),
  File "/home/pi/venvs/smartgic-api/lib/python3.7/site-packages/soco/music_services/music_service.py", line 206, in call
    result_elt = message.call()
  File "/home/pi/venvs/smartgic-api/lib/python3.7/site-packages/soco/soap.py", line 320, in call
    raise SoapFault(faultcode, faultstring, faultdetail)
soco.soap.SoapFault: ns8eeLx:Client.TokenRefreshRequired: tokenRefreshRequired
```

pwt commented:
> 2. I'm quite interested in that fact that the play test failed and in whether that is a problem with the test or the music service. Can I persuade you to test separately, whether you can add e.g. a track that you got from the Spotify service to the queue and play it?

Yes, I can add a Spotify track to the queue using a Sonos controller app, and play it.

I also ran a simple SoCo test to search for and add a single track to the queue, using the latest commit (https://github.com/SoCo/SoCo/pull/763/commits/75fc273ea43482709738288d5137f210e63ba00f). Searching works, but adding to the queue does not, as before. In case it's useful, the DEBUG log is at: https://gist.github.com/pwt/bc47945664ed039ae35f4c3489084510 (with the login token redacted).

pwt commented:
> I have pushed a new commit which tries to handle the token refresh by parsing the new private key and token passed along with the error when the old one has expired. Can you test if this works for you. I.e. pull the new commit, run any call on the music service requiring auth (e.g. search for a track) and check your tokenstore before and after to see if it was properly updated.

I hit `TokenRefreshRequired` with Spotify, but the token store didn't get updated (using commit https://github.com/SoCo/SoCo/commit/75fc273ea43482709738288d5137f210e63ba00f).

amelchio commented:
It seems like something still needs to be pushed, the last commit just adds a class definition.

An issue for me is that the use of `discovery.any_soco()` in `music_service.py` means that I have to unplug my network cable to be able to use this feature (my Sonos is only available on my wifi interface). Being able to provide my own `SoCo` object would be helpful.

Also note that we now have support for multi-household so the configuration file should probably be keyed by household ID because each household might use different music service accounts.

goldyfruit commented:
I tried with Amazon Music, `part1` is working fine _(I got the URLand used it)_ but I guess the code is not good then because of that `part2` failed.

`part1` return this:
```
('https://www.amazon.com/ap/oa?client_id=amzn1.application-oa2-client.5908e9bab8984f9caadb5a289687c4a5&scope=amazon_music%3Aaccess&response_type=code&state=XXXXXXXXXXX%3D&redirect_uri=https%3A%2F%2Fsonos-na.amazon.com%2Flwa', 'blablablabla12345=')
```

goldyfruit commented:
> It seems like something still needs to be pushed, the last commit just adds a class definition.
Same tought for me. :+1:

goldyfruit commented:
@KennethNielsen happy new year 2021!
Do you think you will be able to push your missing changes _(don't worry I know time is hard to find)_ to let us test the token refresh?

KennethNielsen commented:
@pwt @amelchio @goldyfruit I'm sorry. Yes, I forgot to push the change. I'm hoping to pick this up soon, fix the tests also etc. Can you let me know if it works for you know with the token refresh?

@amelchio great idea using the hoursehold id in the configuration and being able to pass a soco object. I will look into that.

goldyfruit commented:
@KennethNielsen I'll try this week!

pwt commented:
Reauthentication is working for me with the current version (336ef72) and Spotify, although I get occasional token expiry exceptions that self-resolve on the next run.

I'm still getting errors for each `add_to_queue` call, as [above](https://github.com/SoCo/SoCo/pull/763#issuecomment-736602271). In case it's useful, if I inspect what's being sent to the player, I get the following ... note the `'can_play' = False`.

![screenshot_213](https://user-images.githubusercontent.com/1089749/106872894-bff79480-66cb-11eb-981c-2a0a973d8ae3.png)

goldyfruit commented:
@KennethNielsen The token refresh worked for me! 
```
SoapFault
ns8eeLx:Client.TokenRefreshRequired
<Element 'detail' at 0x7f8ea22e08>
<?xml version="1.0" ?>
<detail xmlns:ms="http://www.sonos.com/Services/1.1">
  <refreshAuthTokenResult>
    <ms:authToken>XXX</ms:authToken>
    <ms:privateKey>YYY</ms:privateKey>
    <ms:userInfo>
      <ms:userIdHashCode>1234567890qbcdefgh</ms:userIdHashCode>
      <ms:accountTier>paidPremium</ms:accountTier>
      <ms:nickname>Gaëtan Trellu</ms:nickname>
    </ms:userInfo>
  </refreshAuthTokenResult>
</detail>

##############################REAUTH TO##############################
##############################REAUTH TO##############################
##############################REAUTH TO##############################
##############################REAUTH TO##############################
##############################REAUTH TO##############################
##############################REAUTH TO##############################
##############################REAUTH TO##############################
##############################REAUTH TO##############################
##############################REAUTH TO##############################
##############################REAUTH TO##############################
ZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ
```

KennethNielsen commented:
@pwt @goldyfruit thanks so much for testing. I have to honest here and say that I have a **really** tight work deadline, which means that I will not be able to get back to this until week 9.

goldyfruit commented:
> @pwt @goldyfruit thanks so much for testing. I have to honest here and say that I have a **really** tight work deadline, which means that I will not be able to get back to this until week 9.

And it's totally fine by me :)

goldyfruit commented:
So of course... I had to post my previous message to see this error during the SoCo token refresh.
```
Received {'Accept': 'text/xml', 'Content-Type': 'text/xml;charset=UTF-8', 'SOAPAction': '""', 'Content-Length': '430', 'Expires': 'Wed, 10 Mar 2021 18:49:54 GMT', 'Cache-Control': 'max-age=0, no-cache, no-store', 'Pragma': 'no-cache', 'Date': 'Wed, 10 Mar 2021 18:49:54 GMT', 'Connection': 'close'}, <?xml version='1.0' encoding='UTF-8'?><soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/"><soapenv:Header/><soapenv:Body><soapenv:Fault><faultcode>soapenv:Server</faultcode><faultstring xml:lang="en">Spotify service unavailable</faultstring><detail><ExceptionInfo>Spotify service is currently unavailable.</ExceptionInfo><SonosError>3</SonosError></detail></soapenv:Fault></soapenv:Body></soapenv:Envelope>
SoapFault
soapenv:Server
<Element 'detail' at 0x7f9cf50d18>
Other Fault
soapenv:Server
<Element 'detail' at 0x7f9cf50d18>
```
```
SoapFault
soapenv:Server
<Element 'detail' at 0x7f97d2a8b8>
Other Fault
soapenv:Server
<Element 'detail' at 0x7f97d2a8b8>
Traceback (most recent call last):
  File "/home/pi/.local/lib/python3.7/site-packages/soco/music_services/music_service.py", line 207, in call
    result_elt = message.call()
  File "/home/pi/.local/lib/python3.7/site-packages/soco/soap.py", line 320, in call
    raise SoapFault(faultcode, faultstring, faultdetail)
soco.soap.SoapFault: soapenv:Server: Spotify service unavailable

The above exception was the direct cause of the following exception:
```

goldyfruit commented:
After few minutes it's works again, it feels like what @pwt experienced.

tipofthesowrd commented:
I thought you'd like to know I just tried running this on my Windows machine and was able to add a track from Deezer into my Sonos Queue. Since my subscription and authentication was active I didn't have to reauthenticate

goldyfruit commented:
> I thought you'd like to know I just tried running this on my Windows machine and was able to add a track from Deezer into my Sonos Queue. Since my subscription and authentication was active I didn't have to reauthenticate

Did you do anything special to make it work? I registered the service via the application but when I tried to use it via SoCo I got this error.
```
soco.exceptions.MusicServiceException: No account found for service: 'Deezer'
```

goldyfruit commented:
I confirm, Deezer is working but OMG the URL is sooooooo long!

tipofthesowrd commented:
I noticed some strange behaviour with Deezer. I'm not sure it's related to SoCo or Sonos itself.

Maybe somebody can some ideas why adding a single playlist via add_to_queue() is not possible via get_metadata().
(It returns 100 tracks).

however, adding a 100 track playlist from a list of playlists via add_to_queue works fine

```python
    player = SoCo("192.168.2.131")
    deezer = MusicService("Deezer")

    # This is superfast
    player.clear_queue()
    sleep(1)
    results = deezer.get_metadata("playlists")
    player.add_to_queue(choice(results))
    player.play()
    sleep(20)
    player.stop()

    # This doesn't work :-(
    player.clear_queue()
    sleep(1)
    results = deezer.get_metadata("playlist-6429899564")
    print("Got {} items ".format(len(results)))
    player.add_to_queue(results)

    # works but so slow it crashes Sonos player
    player.clear_queue()
    sleep(1)
    results = deezer.get_metadata("playlist-6429899564")
    print("Got {} items ".format(len(results)))
    player.add_multiple_to_queue(results)
```

When trying to add_to_queue based on one playlist I get the following Error:

```
  File "/root/pysonos/examples/commandline/deezer.py", line 103, in <module>
    player.add_to_queue(results)
  File "/usr/local/lib/python3.8/site-packages/pysonos/core.py", line 129, in inner_function
    return function(self, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/pysonos/core.py", line 1993, in add_to_queue
    metadata = to_didl_string(queueable_item)
  File "/usr/local/lib/python3.8/site-packages/pysonos/data_structures.py", line 75, in to_didl_string
    didl.append(arg.to_element())
AttributeError: 'SearchResult' object has no attribute 'to_element'
```

goldyfruit commented:
@tipofthesowrd `add_to_queue()` can only take one playlist, try this _(I used your code)_:

```
from soco.music_services import MusicService
from soco.discovery import by_name
from random import choice

provider = MusicService('Deezer')

player = by_name("Library")
player.clear_queue()
results = provider.get_metadata("playlist-6429899564")
print("Got {} items ".format(len(results)))
player.add_to_queue(choice(results))
```

tipofthesowrd commented:
Yes, that works, since choice() is only picking one item of the playlist it only adds one track of that playlist, not the entire playlist. 

```python
    results = deezer.get_metadata("playlist-6429899564")
    print("Got {} items in the playlist".format(len(results)))
    player.add_to_queue(choice(results))
    print("Got {} items in queue".format(len(player.get_queue())))
```

goldyfruit commented:
This is how I handle authentication for Spotify and Deezer, I'm using a private URL shortener system.
You could have a look here: https://github.com/smartgic/mycroft-sonos-controller-skill/blob/20.8.1/utils.py#L15-L66

deekue commented:
Has anyone had any luck using this with YouTube Music?
I'm getting a 403 "The request is missing a valid API key."

I don't see any mention of that key in the [Sonos Music API docs](https://developer.sonos.com/reference/sonos-music-api/getapplink/) though

```
>>> MusicService.get_data_for_name('YouTube Music')
{'Id': '284', 'Name': 'YouTube Music', 'Version': '1.1', 'Uri': 'https://music.googleapis.com/v1:sendRequest', 'SecureUri': 'https://music.googleapis.com/v1:sendRequest', 'ContainerType': 'MService', 'Capabilities': '32473609', 'Auth': 'AppLink', 'PollInterval': '900', 'ManifestUri': 'https://mnfts.ws.sonos.com/p/ytm.11.json', 'ServiceID': '284', 'ServiceType': '72711'}
>>> ytm = MusicService('YouTube Music')
>>> link = ytm.device_or_app_link_auth_part1()
Traceback (most recent call last):
  File "test_ms_auth.py", line 7, in <module>
    reg_url, link_code = music_service.device_or_app_link_auth_part1()
  File "/home/apathy/src/kennethnielsen-soco/soco/music_services/music_service.py", line 804, in device_or_app_link_auth_part1
    return self.soap_client.device_or_app_link_auth_part1()
  File "/home/apathy/src/kennethnielsen-soco/soco/music_services/music_service.py", line 294, in device_or_app_link_auth_part1
    result = self.call("getAppLink", [("householdId", self._household_id)])[
  File "/home/apathy/src/kennethnielsen-soco/soco/music_services/music_service.py", line 210, in call
    result_elt = message.call()
  File "/home/apathy/src/kennethnielsen-soco/soco/soap.py", line 324, in call
    response.raise_for_status()
  File "/usr/lib/python3/dist-packages/requests/models.py", line 940, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://music.googleapis.com/v1:sendRequest
```

if I try with logging level DEBUG
```
DEBUG:soco.soap:Sending {'Content-Type': 'text/xml; charset="utf-8"', 'SOAPACTION': '"http://www.sonos.com/Services/1.1#getAppLink"', 'Accept-Encoding': 'gzip, deflate', 'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:88.0) Gecko/20100101 Firefox/88.0', 'Accept-Language': 'en-US'}, <?xml version="1.0" ?>
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
  <s:Header>
    <credentials xmlns="http://www.sonos.com/Services/1.1">
      <deviceId>00-0E-58-BB-54-C8:F</deviceId>
      <deviceProvider>Sonos</deviceProvider>
      <context>
        <timezone>+01:00</timezone>
      </context>
      <loginToken>
        <householdId>Sonos_K1GlMV3xzCk1SyTO3kYjsdNCQb</householdId>
      </loginToken>
    </credentials>
  </s:Header>
  <s:Body>
    <getAppLink xmlns="http://www.sonos.com/Services/1.1">
      <householdId>Sonos_K1GlMV3xzCk1SyTO3kYjsdNCQb</householdId>
    </getAppLink>
  </s:Body>
</s:Envelope>

DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): music.googleapis.com:443
DEBUG:urllib3.connectionpool:https://music.googleapis.com:443 "POST /v1:sendRequest HTTP/1.1" 403 None
DEBUG:soco.soap:Received {'Vary': 'Origin, X-Origin, Referer', 'Content-Type': 'application/json; charset=UTF-8', 'Content-Encoding': 'gzip', 'Date': 'Sat, 05 Jun 2021 17:16:37 GMT', 'Server': 'ESF', 'Cache-Control': 'private', 'X-XSS-Protection': '0', 'X-Frame-Options': 'SAMEORIGIN', 'X-Content-Type-Options': 'nosniff', 'Alt-Svc': 'h3-29=":443"; ma=2592000,h3-T051=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"', 'Transfer-Encoding': 'chunked'}, {
  "error": {
    "code": 403,
    "message": "The request is missing a valid API key.",
    "status": "PERMISSION_DENIED"
  }
}

Traceback (most recent call last):
  File "<stdin>", line 11, in <module>
  File "/home/apathy/src/kennethnielsen-soco/soco/music_services/music_service.py", line 806, in device_or_app_link_auth_part1
    return self.soap_client.device_or_app_link_auth_part1()
  File "/home/apathy/src/kennethnielsen-soco/soco/music_services/music_service.py", line 296, in device_or_app_link_auth_part1
    result = self.call("getAppLink", [("householdId", self._household_id)])[
  File "/home/apathy/src/kennethnielsen-soco/soco/music_services/music_service.py", line 212, in call
    result_elt = message.call()
  File "/home/apathy/src/kennethnielsen-soco/soco/soap.py", line 324, in call
    response.raise_for_status()
  File "/usr/lib/python3/dist-packages/requests/models.py", line 940, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://music.googleapis.com/v1:sendRequest

[Done] exited with code=1 in 0.621517 seconds
```

svrooij commented:
I’m no python programmer but I’ve done some reverse engineering of how the music services work in Sonos.

all documented [here](https://svrooij.io/sonos-api-docs/music-services.html#authentication) as part of my let’s generate all the services from the Sonos device discovery files.

## Suggested change > Save tokens and keys in Sonos device

Instead of using a JSON to store the tokens, why don’t we specify how we (as open-source Sonos libraries) will be saving these tokens in the Sonos system.

The SystemPropertiesService has 2 actions that let you save a string in the Sonos speaker, and load that same string. As documented here [GetString](https://svrooij.io/sonos-api-docs/services/system-properties.html#getstring) and [SetString](https://svrooij.io/sonos-api-docs/services/system-properties.html#setstring).

In my [node library for Sonos](https://svrooij.io/node-sonos-ts/), I save a list of logged on music service ids as a string of numbers separated by the `|` char. That way every other app could now which services you subscribed to.

keys im using:

- `sonos-ts-accounts` for all the accounts logged in to like `2|9` for deezer and Spotify.
- `sonos-ts-9-key` for the key for Spotify
- `sonos-ts-9-token` for the token for Spotify

Ps. This is just a suggestion, I won’t be building it because I lack python skills. Let’s all work together to make controlling Sonos speakers better for everyone.

KennethNielsen commented:
> After few minutes it's works again, it feels like what @pwt experienced.

Ok. I think we will tackle this later. It may simply be that the content servers are unstable and we need some sort of retrying going on and every other client of that music service worked around it or (which is somewhat more likely) we are using it wrong, maybe triggering a rate limit of what do I know. In any case, I suggest to postpone that particular problem, get this PR merged, so everyone isn't held back by it.

KennethNielsen commented:
@deekue is that using this code in this PR?

KennethNielsen commented:
@svrooij hey, yeah I saw you did that. I've been looking quite a bit on your docs and code. I don't like the idea much, as I don't like storing authentication information where it is readily available on the local network, so, at least for now, it will be saved in the text file, but with the option for anyone to add another way to store tokens.

svrooij commented:
Personally I hate the fact that Sonos has removed this information 😏

Just use what works. You could also encrypt them before storing and only saving the encryption key in the json file. That way you can still share the tokens between apps.

goldyfruit commented:
Just try to load @KennethNielsen branch and now I get this error:
```
  File "/opt/mycroft-venv/lib/python3.7/site-packages/soco/__init__.py", line 12, in <module>
    from .core import SoCo
  File "/opt/mycroft-venv/lib/python3.7/site-packages/soco/core.py", line 30, in <module>
    from .data_structures_entry import from_didl_string
  File "/opt/mycroft-venv/lib/python3.7/site-packages/soco/data_structures_entry.py", line 13, in <module>
    from .music_services.data_structures import get_class
  File "/opt/mycroft-venv/lib/python3.7/site-packages/soco/music_services/__init__.py", line 5, in <module>
    from .music_service import MusicService
  File "/opt/mycroft-venv/lib/python3.7/site-packages/soco/music_services/music_service.py", line 20, in <module>
    from ..compat import parse_qs, quote_url, urlparse
ModuleNotFoundError: No module named 'soco.compat'
```
Any idea of what could be the issue?

goldyfruit commented:
Does this PR has been replaced by #847 ?

goldyfruit commented:
> > Does this PR has been replaced by #847 ?
> 
> No. The sharelink plugin PRs (#838, #848, etc.) are convenient but not a replacement for full music service support (ability to search. etc.).

Thanks for the clarification.

goldyfruit commented:
> Has anyone had any luck using this with YouTube Music? I'm getting a 403 "The request is missing a valid API key."
> 
> I don't see any mention of that key in the [Sonos Music API docs](https://developer.sonos.com/reference/sonos-music-api/getapplink/) though
> 
> ```
> >>> MusicService.get_data_for_name('YouTube Music')
> {'Id': '284', 'Name': 'YouTube Music', 'Version': '1.1', 'Uri': 'https://music.googleapis.com/v1:sendRequest', 'SecureUri': 'https://music.googleapis.com/v1:sendRequest', 'ContainerType': 'MService', 'Capabilities': '32473609', 'Auth': 'AppLink', 'PollInterval': '900', 'ManifestUri': 'https://mnfts.ws.sonos.com/p/ytm.11.json', 'ServiceID': '284', 'ServiceType': '72711'}
> >>> ytm = MusicService('YouTube Music')
> >>> link = ytm.device_or_app_link_auth_part1()
> Traceback (most recent call last):
>   File "test_ms_auth.py", line 7, in <module>
>     reg_url, link_code = music_service.device_or_app_link_auth_part1()
>   File "/home/apathy/src/kennethnielsen-soco/soco/music_services/music_service.py", line 804, in device_or_app_link_auth_part1
>     return self.soap_client.device_or_app_link_auth_part1()
>   File "/home/apathy/src/kennethnielsen-soco/soco/music_services/music_service.py", line 294, in device_or_app_link_auth_part1
>     result = self.call("getAppLink", [("householdId", self._household_id)])[
>   File "/home/apathy/src/kennethnielsen-soco/soco/music_services/music_service.py", line 210, in call
>     result_elt = message.call()
>   File "/home/apathy/src/kennethnielsen-soco/soco/soap.py", line 324, in call
>     response.raise_for_status()
>   File "/usr/lib/python3/dist-packages/requests/models.py", line 940, in raise_for_status
>     raise HTTPError(http_error_msg, response=self)
> requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://music.googleapis.com/v1:sendRequest
> ```
> 
> if I try with logging level DEBUG
> 
> ```
> DEBUG:soco.soap:Sending {'Content-Type': 'text/xml; charset="utf-8"', 'SOAPACTION': '"http://www.sonos.com/Services/1.1#getAppLink"', 'Accept-Encoding': 'gzip, deflate', 'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:88.0) Gecko/20100101 Firefox/88.0', 'Accept-Language': 'en-US'}, <?xml version="1.0" ?>
> <s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
>   <s:Header>
>     <credentials xmlns="http://www.sonos.com/Services/1.1">
>       <deviceId>00-0E-58-BB-54-C8:F</deviceId>
>       <deviceProvider>Sonos</deviceProvider>
>       <context>
>         <timezone>+01:00</timezone>
>       </context>
>       <loginToken>
>         <householdId>Sonos_K1GlMV3xzCk1SyTO3kYjsdNCQb</householdId>
>       </loginToken>
>     </credentials>
>   </s:Header>
>   <s:Body>
>     <getAppLink xmlns="http://www.sonos.com/Services/1.1">
>       <householdId>Sonos_K1GlMV3xzCk1SyTO3kYjsdNCQb</householdId>
>     </getAppLink>
>   </s:Body>
> </s:Envelope>
> 
> DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): music.googleapis.com:443
> DEBUG:urllib3.connectionpool:https://music.googleapis.com:443 "POST /v1:sendRequest HTTP/1.1" 403 None
> DEBUG:soco.soap:Received {'Vary': 'Origin, X-Origin, Referer', 'Content-Type': 'application/json; charset=UTF-8', 'Content-Encoding': 'gzip', 'Date': 'Sat, 05 Jun 2021 17:16:37 GMT', 'Server': 'ESF', 'Cache-Control': 'private', 'X-XSS-Protection': '0', 'X-Frame-Options': 'SAMEORIGIN', 'X-Content-Type-Options': 'nosniff', 'Alt-Svc': 'h3-29=":443"; ma=2592000,h3-T051=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"', 'Transfer-Encoding': 'chunked'}, {
>   "error": {
>     "code": 403,
>     "message": "The request is missing a valid API key.",
>     "status": "PERMISSION_DENIED"
>   }
> }
> 
> Traceback (most recent call last):
>   File "<stdin>", line 11, in <module>
>   File "/home/apathy/src/kennethnielsen-soco/soco/music_services/music_service.py", line 806, in device_or_app_link_auth_part1
>     return self.soap_client.device_or_app_link_auth_part1()
>   File "/home/apathy/src/kennethnielsen-soco/soco/music_services/music_service.py", line 296, in device_or_app_link_auth_part1
>     result = self.call("getAppLink", [("householdId", self._household_id)])[
>   File "/home/apathy/src/kennethnielsen-soco/soco/music_services/music_service.py", line 212, in call
>     result_elt = message.call()
>   File "/home/apathy/src/kennethnielsen-soco/soco/soap.py", line 324, in call
>     response.raise_for_status()
>   File "/usr/lib/python3/dist-packages/requests/models.py", line 940, in raise_for_status
>     raise HTTPError(http_error_msg, response=self)
> requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://music.googleapis.com/v1:sendRequest
> 
> [Done] exited with code=1 in 0.621517 seconds
> ```

Have you been successful with this service?

goldyfruit commented:
For Amazon Music, it seems that in `device_or_app_link_auth_part2` the `self._device_id` is not retrieve as expected which male this call invalid.
```python
        result = self.call(
            "getDeviceAuthToken",
            [
                ("householdId", self._household_id),
                ("linkCode", link_code),
                ("linkDeviceId", self._device_id),
            ],
        )["getDeviceAuthTokenResult"]
```

goldyfruit commented:
I was finally able to get a token from Amazon Music, the problem was due to the `linkDeviceId` that was send to Amazon, the value should not be the one from `self._device_id` but `linkDeviceId` returned during `device_or_app_link_auth_part1`.

jjlawren commented:
([Comment](https://github.com/SoCo/SoCo/pull/869#issuecomment-991353397) copied from #869)

I did a bit of testing with this against a [Navidrome](https://www.navidrome.org) + [bonob](https://github.com/simojenki/bonob) setup. I was able to link & authorize properly, but searching did not work. It seems that the service provides categories but not associated `mappedId` values in the `presentationMap.xml`:
```xml
<PresentationMap type="Search">
  <Match>
    <SearchCategories>
      <Category id="artists"/>
      <Category id="albums"/>
      <Category id="tracks"/>
    </SearchCategories>
  </Match>
</PresentationMap>
```
It turns out that using the `id` itself as-is works fine with this service. Providing this as a fallback fixed my issue:
```diff
diff --git a/soco/music_services/music_service.py b/soco/music_services/music_service.py
index 53f9ab7..b397143 100644
--- a/soco/music_services/music_service.py
+++ b/soco/music_services/music_service.py
@@ -655,7 +655,7 @@ class MusicService:
         if categories is None:
             return self._search_prefix_map
         for cat in categories:
-            self._search_prefix_map[cat.get("id")] = cat.get("mappedId")
+            self._search_prefix_map[cat.get("id")] = cat.get("mappedId") or cat.get("id")
         custom_categories = pmap_root.findall(".//SearchCategories/CustomCategory")
         for cat in custom_categories:
             self._search_prefix_map[cat.get("stringId")] = cat.get("mappedId")
```

KennethNielsen commented:
Superseded by #883, all future discussion should take place there. Thanks for all the feedback and testing.
